### ISSUE TITLE
Split GetX into smaller components

### ISSUE THREAD

**Issue by rockerhieu** [jonataslaw/getx#288]
**Is your feature request related to a problem? Please describe.**
Currently GetX is like a Swiss Army knife whose too many responsibilities and features:
* Navigation
* Dependencies injection
* State management
* Reactive state management
* Other utilities that doesn't require `context`

While it's nice to have many features, I find it's difficult to maintain and only use a subset of the features.

**Describe the solution you'd like**
I think it will make more sense if you break GetX into smaller components.

**Comment by jonataslaw** [comment]
Hi, they are already divided into components, and only the resources you use are initialized.
You can use Dart inspect to see what's in your app's memory.
If you use only routes, only route components are placed in memory. If you use only the state manager, only state managers are called. I will put this explicitly in the readme, because it seems that it is a very common mistake for those who see the package, to think that there will be some blot in your app.

**Comment by rockerhieu** [comment]
I meant GetX should be divided into smaller libraries. Let's say I want to use only `getx-navigation`, then I should not see other unrelated extension methods in `Get`, and also not `.obs`.

These libraries are isolated and hence easier to maintain and less bugs. If I choose to use `getx-navigation` I will not getting any bugs in `getx-bindings`, nor `getx-state`, nor `getx-state-rx`, ... And another benefit is it will be easier to read and understand documentation because users will only need to learn one library at a time.

**Comment by stefandevo** [comment]
I understand your concern, but I believe for example bindings are executed when using the navigation part, state is used by the bindings... so internally most of these things are using each other. Of course they can all start referencing each other,and in the end it gets more complex for everyone, and you don't solve the issue where there is a bug in one place it won't affect the other package...
Bottom line; I think some of the features will always be in the same library. And BTW, just use it ALL because it's all freaky awesome!

**Comment by jonataslaw** [comment]
I think one problem that this package tries to solve is that you also have 100 dependencies in pubspec. Here is a printout of a project of mine after updating Flutter 1.9.
Everything broke. All of my libraries broke down, and I lost 2 days of work just looking for errors to know where they came from and what lib they came from.
I confess that it was one of the most stressful situations I have ever gone through in my life. It was exactly on that day that I decided to do everything in one package.
I know this looks a lot like the package being based on my personal experiences, but I am a programmer, and I try to solve problems always from the programmer's perspective. I don't care about anything other than making my life and other devs easier with this library. Every time I go through a frustrating experience, I write it down in my schedule, and try to resolve it after completing the project.
And you will use state, route and dependency management in your project, this is the basis for any project. You can completely isolate state management, for example, but you will lose the benefits of integrating them.

![96130029_581336076096723_2612998381366149120_n](https://user-images.githubusercontent.com/35742643/85869309-e488a700-b7a1-11ea-8bd2-5d20e8964559.jpg)

**Comment by rockerhieu** [comment]
> I understand your concern, but I believe for example bindings are executed when using the navigation part, state is used by the bindings... so internally most of these things are using each other. Of course they can all start referencing each other,and in the end it gets more complex for everyone, and you don't solve the issue where there is a bug in one place it won't affect the other package...
> Bottom line; I think some of the features will always be in the same library. And BTW, just use it ALL because it's all freaky awesome!

I think that's exactly the issue I have mentioned, because everything is tight decoupling together as a whole right now. Separating into smaller libraries will introduce separation of concerns hence more robust design.

If you can use all features of GetX then that's good for you. Maybe it's just me that I'm finding it's hard to put too many eggs into one basket. I like GetX because of its elegant API design, but only my first day with GetX and I already got 2 critical bugs: #287 #293.

**Comment by jonataslaw** [comment]
> > I understand your concern, but I believe for example bindings are executed when using the navigation part, state is used by the bindings... so internally most of these things are using each other. Of course they can all start referencing each other,and in the end it gets more complex for everyone, and you don't solve the issue where there is a bug in one place it won't affect the other package...
> > Bottom line; I think some of the features will always be in the same library. And BTW, just use it ALL because it's all freaky awesome!
> 
> I think that's exactly the issue I have mentioned, because everything is tight coupling together as a whole right now. Separating into smaller libraries will introduce separation of concerns hence more robust design.
> 
> If you can use all features of GetX then that's good for you. Maybe it's just me that I'm finding it's hard to put too many eggs into one basket. I like GetX because of its elegant API design, but only my first day with GetX and I already got 2 critical bugs: #287 #293.

One of the issues is not a bug, it was just inappropriate use of the resource, there are some closed issues that dealt with the same subject, if you had looked for it, it would not be necessary to open a new issue, but I am inserting it in the documentation to make it clear about the types of Get.lazyPut.
The second problem lasted 24 hours in the air, was introduced by an update, and that only happens in one version, in a very specific case. It has been fixed since yesterday.
